### DISCUSSION TITLE
_clean_tis_without_dagrun run long time

### DISCUSSION THREAD

**Discussion by zqd759592942** [apache/airflow#30383] [Q&A]
airflow version 2.0.2 

![image](https://user-images.githubusercontent.com/37291301/229056770-bc2406b8-1eca-4c3e-a6cd-ba1b156ba67c.png)

![image](https://user-images.githubusercontent.com/37291301/229057472-46677c84-e8bc-46db-8f5c-aff0c92380dd.png)

airlfow  scheduling cycle time is too long.
The final localization is caused by the full table scan of the 【_clean_tis_without_dagrun】 method,It is performed every 15 seconds by default.
I learned from the source code that the latest version seems to solve this problem,However, the need to change the table structure is too expensive, our very important formal environment can not be modified casually, I want to ask, is there any other way to solve this problem? 
Thank you very much!

**Comment by potiuk** [comment]
You could use `airflow db clean` if you had it and remove old records. But your version of Airlfow does not yet have the `clean` method - look at the code what it does and perform similar cleanup via a custom SQL script doing the same for your version (trimmed down to 2.0.2 as your db structure is different). 

There were also some https://github.com/teamclairvoyant/airflow-maintenance-dags/blob/master/db-cleanup/README.md that people used in the past before db clean, but I cannot really tell what's the quality and applicability of those are to your version.

**Reply by zqd759592942** [reply]
Thanks to your suggestion, we currently have 10,000 DAG files, about six or seven million tasks run every month, and now the data saving strategy is to clean up once a month, only retain the data from a month ago to the present, we try to reduce the amount of data saved

**Comment by potiuk** [comment]
BTW. SERIOUSLY consider upgrading to latest released version (we are **just** voting for 2.5.3rc2). There are literally thousands of fixes since 2.0.2 and you are not only missing on those but also there were quite a number of security vulnerabilities since 2.0.2 so you are also exposed to some secuirty threats by not upgrading.

**Reply by zqd759592942** [reply]
Okay, we'll try to push Airflow's version iteration and look forward to some new features
